config/v1/types_oauth.go

```diff
@@ -1,37 +1,420 @@
-// TODO this object is an example of a possible grouping and is subject to change or removal
 type OAuth struct {
-	metav1.TypeMeta `json:",inline"`
-	// Standard object's metadata.
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	// spec holds user settable values for configuration
-	Spec OAuthSpec `json:"spec"`
-	// status holds observed values from the cluster. They may not be overridden.
-	Status OAuthStatus `json:"status"`
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              OAuthSpec   `json:"spec"`
+	Status            OAuthStatus `json:"status,omitempty"`
 }
 
 type OAuthSpec struct {
-	// options for configuring the embedded oauth server.
-	// possibly wellknown?
+	//IdentityProviders is an ordered list of ways for a user to identify themselves
+	IdentityProviders []IdentityProvider `json:"identityProviders"`
+
+	// GrantConfig describes how to handle grants
+	GrantConfig GrantConfig `json:"grantConfig"`
+
+	// TokenConfig contains options for authorization and access tokens
+	TokenConfig TokenConfig `json:"tokenConfig"`
+
+	// Templates allow you to customize pages like the login page.
+	Templates *OAuthTemplates `json:"templates"`
 }
 
 type OAuthStatus struct {
+	operatorsv1alpha1api.OperatorStatus `json:",inline"`
+}
+
+// OAuthTemplates allow for customization of pages like the login page
+type OAuthTemplates struct {
+	// Login is a path to a file containing a go template used to render the login page.
+	// If unspecified, the default login page is used.
+	Login string `json:"login"`
+
+	// ProviderSelection is a path to a file containing a go template used to render the provider selection page.
+	// If unspecified, the default provider selection page is used.
+	ProviderSelection string `json:"providerSelection"`
+
+	// Error is a path to a file containing a go template used to render error pages during the authentication or grant flow
+	// If unspecified, the default error page is used.
+	Error string `json:"error"`
+}
+
+// GrantHandlerType are the valid strategies for handling grant requests
+type GrantHandlerType string
+
+const (
+	// GrantHandlerAuto auto-approves client authorization grant requests
+	GrantHandlerAuto GrantHandlerType = "auto"
+	// GrantHandlerPrompt prompts the user to approve new client authorization grant requests
+	GrantHandlerPrompt GrantHandlerType = "prompt"
+	// GrantHandlerDeny auto-denies client authorization grant requests
+	GrantHandlerDeny GrantHandlerType = "deny"
+)
+
+// GrantConfig holds the necessary configuration options for grant handlers
+type GrantConfig struct {
+	// Method: allow, deny, prompt
+	Method GrantHandlerType
+
+	// ServiceAccountMethod is used for determining client authorization for service account oauth client.
+	// It must be either: deny, prompt
+	ServiceAccountMethod GrantHandlerType
+}
+
+var ValidGrantHandlerTypes = sets.NewString(string(GrantHandlerAuto), string(GrantHandlerPrompt), string(GrantHandlerDeny))
```

Comment:
We only place the types, constants, and our schemes in openshift/api.  We don't place validation related items here.

Reply:
removed in 88e7c4e